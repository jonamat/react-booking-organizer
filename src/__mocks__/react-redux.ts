import { Customer, Employee, Reservation, Service } from '../types';

export const __mockCustomers: Array<Customer> = [
    {
        id: 'customer-FAKE-ID1',
        name: 'Customer name',
        surname: 'Surname',
    },
    {
        id: 'customer-FAKE-ID2',
        name: 'Other name',
        surname: 'Other surname',
    },
];

export const __mockEmployees: Array<Employee> = [
    {
        id: 'employee-FAKE-ID1',
        name: 'Employee name 1',
    },
    {
        id: 'employee-FAKE-ID2',
        name: 'Employee name 2',
    },
];

export const __mockServices: Array<Service> = [
    {
        id: 'service-FAKE-ID1',
        name: 'fake service 1',
        averageDuration: 30,
    },
    {
        id: 'service-FAKE-ID2',
        name: 'fake service 2',
        averageDuration: 45,
    },
];

export const __mockReservations: Array<Reservation> = [
    {
        id: 'reservation-FAKE-ID1',
        date: new Date('1 Jan 2000 10:00 GMT+0100'),
        serviceId: __mockServices[0].id,
        customerId: __mockCustomers[0].id,
        employeeId: __mockEmployees[0].id,
        notes: 'something',
    },
    {
        id: 'reservation-FAKE-ID2',
        date: new Date('1 Jan 2040 10:00 GMT+0100'),
        serviceId: __mockServices[1].id,
        customerId: __mockCustomers[0].id,
        employeeId: __mockEmployees[1].id,
        notes: 'something2',
    },
];

export let useStore = () => ({
    getState: () => ({
        database: {
            customers: __mockCustomers,
            employees: __mockEmployees,
            services: __mockServices,
            reservations: __mockReservations,
        },
    }),
});

export const __setUseStore = (value) => (useStore = value);
export const useDispatch = () => undefined;
export const connect = () => (component) => component;
