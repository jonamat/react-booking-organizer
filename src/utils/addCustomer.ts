import firebase from 'firebase/app';

import { NewCustomer, isNewCustomer, Customer } from '../types';
import { CUSTOMERS_COLLECTION } from '../config';

const REQUIRED_FIELDS = ['name'];

/**
 * Validate user inputs and send them to database
 * @param formData Raw data from form
 * @param birthday Birthday Date instance
 */
const addCustomer = async (formData: FormData, birthday: Date | null): Promise<Customer> => {
    if (REQUIRED_FIELDS.some((fieldName) => !formData.get(fieldName))) throw new Error('Compila i campi obbligatori');

    const newCustomer: any = {
        name: formData.get('name') || undefined,
        surname: formData.get('surname') || undefined,
        additionalIdentifier: formData.get('additionalIdentifier') || undefined,
        phone: formData.get('phone') || undefined,
        birthday: birthday || undefined,
        notes: formData.get('notes') || undefined,
    };

    if (!isNewCustomer(newCustomer)) throw new Error('Errore interno. Contatta il supporto tecnico');

    // Delete empty values
    (Object.keys(newCustomer) as Array<keyof NewCustomer>).forEach((key) => {
        if (newCustomer[key] === undefined) delete newCustomer[key];
    });

    try {
        return {
            id: (await firebase.firestore().collection(CUSTOMERS_COLLECTION).add(newCustomer)).id,
            ...newCustomer,
        };
    } catch (error) {
        console.error(error);
        throw new Error('Connessione al database fallita');
    }
};

export default addCustomer;
