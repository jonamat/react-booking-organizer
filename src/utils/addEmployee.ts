import firebase from 'firebase/app';

import { NewEmployee, Employee, isNewEmployee, WeekWorkingShifts, Holiday } from '../types';
import { EMPLOYEES_COLLECTION } from '../config';

const REQUIRED_FIELDS = ['name'];

/**
 * Validate user inputs and send them to database
 * @param formData Raw data from form
 * @param weekWorkingShifts Valid WeekWorkingShift object
 * @param holidays Valid Holiday array
 */
const addEmployee = async (
    formData: FormData,
    weekWorkingShifts?: WeekWorkingShifts | null,
    holidays?: Array<Holiday> | null,
): Promise<Employee> => {
    if (REQUIRED_FIELDS.some((fieldName) => !formData.get(fieldName))) throw new Error('Compila i campi obbligatori');

    const newEmployee: any = {
        name: formData.get('name') || undefined,
        weekWorkingShifts,
        holidays,
    };

    if (!isNewEmployee(newEmployee)) throw new Error('Errore interno. Contatta il supporto tecnico');

    (Object.keys(newEmployee) as Array<keyof NewEmployee>).forEach((key) => {
        if (newEmployee[key] === undefined) delete newEmployee[key];
    });

    try {
        return {
            id: (await firebase.firestore().collection(EMPLOYEES_COLLECTION).add(newEmployee)).id,
            ...newEmployee,
        };
    } catch (error) {
        console.error(error);
        throw new Error('Connessione al database fallita');
    }
};

export default addEmployee;
