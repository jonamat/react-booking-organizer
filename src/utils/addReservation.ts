import { Employee, Service, Customer, isNewReservation, NewReservation, Reservation } from '../types';
import firebase from 'firebase/app';
import { MaterialUiPickersDate } from '@material-ui/pickers/typings/date';

import { PROGRESSIVE_MINS_WITHIN_HOUR, MIN_STEP, TIME_DATE_FORMAT, RESERVATIONS_COLLECTION } from '../config';
import { isEmployeeBusy, isEmployeeOnDuty } from '.';

const REQUIRED_FIELDS = ['serviceId'];

interface Database {
    employees: Array<Employee>;
    services: Array<Service>;
    reservations: Array<Reservation>;
}

/**
 * Validate user inputs and send them to database
 * @param formData Raw data from form
 * @param customer Customer requesting the reservation
 * @param selectedDate Date of the reservation
 * @param validationData Redux data from store.database for validation
 * @param forceAssignment Force assignment even if employee in not on duty
 */
const addReservation = async (
    formData: FormData,
    customer: Customer | null,
    selectedDate: MaterialUiPickersDate,
    { employees, services, reservations }: Database,
    forceAssignment?: boolean,
): Promise<Reservation> => {
    if (REQUIRED_FIELDS.some((fieldName) => !formData.get(fieldName)) || !customer)
        throw new Error('Compila i campi obbligatori');

    if (!selectedDate || (typeof selectedDate === 'number' && isNaN(selectedDate)))
        throw new Error('Seleziona una data valida');

    if (!PROGRESSIVE_MINS_WITHIN_HOUR.includes(selectedDate.getMinutes()))
        throw new Error(`L'orario indicano non rispetta lo scaglionamento orario dell'agenda.
        Utilizza intervalli di ${MIN_STEP} minuti.`);

    const newReservation: any = {
        date: selectedDate,
        customerId: customer.id,
        serviceId: formData.get('serviceId') || undefined,
        employeeId: formData.get('employeeId') || undefined,
        notes: formData.get('notes') || undefined,
    };

    if (!isNewReservation(newReservation)) throw new Error('Errore interno. Contatta il supporto tecnico');

    const humanReadableDate = newReservation.date.toLocaleDateString('it-IT', TIME_DATE_FORMAT);

    const employee = employees.find((employee) => employee.id === newReservation.employeeId);
    if (!employee) throw new Error('Errore interno - Dipendente non trovato');

    if (isEmployeeBusy(newReservation, services, reservations))
        throw new Error(`${employee?.name} è già impegnato/a con un altro appuntamento ${humanReadableDate}.`);

    if (!forceAssignment && !isEmployeeOnDuty(newReservation.date, employee)) throw new Error('EMPLOYEE_NOT_ON_DUTY');

    (Object.keys(newReservation) as Array<keyof NewReservation>).forEach((key) => {
        if (newReservation[key] === undefined) delete newReservation[key];
    });

    try {
        return {
            id: (await firebase.firestore().collection(RESERVATIONS_COLLECTION).add(newReservation)).id,
            ...newReservation,
        };
    } catch (error) {
        console.error(error);
        throw new Error('Connessione al database fallita');
    }
};

export default addReservation;
