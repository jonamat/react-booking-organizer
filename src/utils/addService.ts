import firebase from 'firebase/app';

import { NewService, isNewService, Service } from '../types';
import { PROGRESSIVE_MINS_WITHIN_MAX_SERVICE, SERVICES_COLLECTION } from '../config';

const REQUIRED_FIELDS = ['name', 'averageDuration'];

/**
 * Validate user inputs and send them to database
 * @param formData Raw data from form
 */
const addService = async (formData: FormData): Promise<Service> => {
    if (REQUIRED_FIELDS.some((fieldName) => !formData.get(fieldName))) throw new Error('Compila i campi obbligatori');

    const newService: any = {
        name: formData.get('name') || undefined,
        averageDuration: parseInt(formData.get('averageDuration') as string) || undefined,
    };

    if (!isNewService(newService) || !PROGRESSIVE_MINS_WITHIN_MAX_SERVICE.includes(newService.averageDuration))
        throw new Error('Errore interno. Contatta il supporto tecnico');

    (Object.keys(newService) as Array<keyof NewService>).forEach((key) => {
        if (newService[key] === undefined) delete newService[key];
    });

    try {
        return {
            id: (await firebase.firestore().collection(SERVICES_COLLECTION).add(newService)).id,
            ...newService,
        };
    } catch (error) {
        console.error(error);
        throw new Error('Connessione al database fallita');
    }
};

export default addService;
