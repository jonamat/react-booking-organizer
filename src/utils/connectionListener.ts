import { isOffline } from '../redux';

/**
 * Trigger a state change on connection changes
 * @param dispatch Redux dispatch function
 */
const connectionListener = (dispatch) => {
    // First check
    dispatch(isOffline(!navigator.onLine));

    window.addEventListener('online', () => dispatch(isOffline(false)));
    window.addEventListener('offline', () => dispatch(isOffline(true)));
};

export default connectionListener;
