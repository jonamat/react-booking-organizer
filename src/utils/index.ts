// Utils
export { default as isEmployeeBusy } from './isEmployeeBusy';
export { default as isEmployeeOnDuty } from './isEmployeeOnDuty';
export { default as generateClosestValidDate } from './generateClosestValidDate';
export { default as getReservationsByDate } from './getReservationsByDate';
export { default as getReservationStatus } from './getReservationStatus';

export { default as formatMinutes } from './formatMinutes';
export { default as hourMinuteToReadableString } from './hourMinuteToReadableString';
export { default as translateWeekDays } from './translateWeekDays';

export { default as databaseSync } from './databaseSync';
export { default as connectionListener } from './connectionListener';

// Auth system
export { default as login } from './login';
export { default as logout } from './logout';

// Delete database elements
export { default as deleteCustomer } from './deleteCustomer';
export { default as deleteReservation } from './deleteReservation';
export { default as deleteEmployee } from './deleteEmployee';
export { default as deleteService } from './deleteService';

// Add database elements
export { default as addCustomer } from './addCustomer';
export { default as addEmployee } from './addEmployee';
export { default as addReservation } from './addReservation';
export { default as addService } from './addService';

// Edit database elements
export { default as editCustomer } from './editCustomer';
export { default as editEmployee } from './editEmployee';
export { default as editReservation } from './editReservation';
export { default as editService } from './editService';
