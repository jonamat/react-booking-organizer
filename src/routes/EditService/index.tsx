import React, { FC, useRef, FormEvent, useState } from 'react';
import {
    Typography,
    TextField,
    Button,
    Divider,
    TextFieldProps,
    FormControl,
    InputLabel,
    Select,
    MenuItem,
    CircularProgress,
} from '@material-ui/core';
import { RouteComponentProps } from 'react-router';
import { toast } from 'react-toastify';
import { useStore } from 'react-redux';

import { editService, formatMinutes } from '../../utils';
import { PROGRESSIVE_MINS_WITHIN_MAX_SERVICE } from '../../config';
import { RootState } from '../../types';
import useStyles from './style';

const AddService: FC<RouteComponentProps<{ serviceId: string }>> = ({ history, match }) => {
    const classes = useStyles();
    const [loading, setLoading] = useState<boolean>(false);
    const inputForm = useRef<HTMLFormElement>(null);
    const {
        database: { services },
    } = useStore<RootState>().getState();

    const fieldProps: TextFieldProps = {
        className: classes.field,
        variant: 'outlined',
        autoComplete: 'off',
    };

    if (!services) throw new Error('Internal error - Resources loading logic broken');

    const serviceId = match.params.serviceId;
    const service = services.find((service) => service.id === serviceId);

    if (!service) {
        toast.error('Servizio non presente nel database');
        history.push('/services');
        return null;
    }

    const handleSubmit = (event: FormEvent<HTMLFormElement>) => {
        event.preventDefault();
        if (!inputForm.current) return;

        // Disable user interactions
        setLoading(true);

        editService(serviceId, new FormData(inputForm.current))
            .then((service) => {
                toast.success('Modificato servizio: ' + service.name);
                history.push('/services');
            })
            .catch((error) => {
                toast.error(error.message);
                setLoading(false);
            });
    };

    return (
        <div className={classes.container} data-testid="edit-service">
            <Typography variant="h6">Aggiungi servizio</Typography>
            <Divider className={classes.divider} />
            <form className={classes.form} ref={inputForm} onSubmit={handleSubmit} data-testid="form">
                <TextField {...fieldProps} name="id" label="ID" required disabled defaultValue={service.id} />
                <TextField
                    {...fieldProps}
                    inputProps={{ 'data-testid': 'name' }}
                    name="name"
                    label="Nome"
                    required
                    defaultValue={service.name}
                />

                <FormControl variant="outlined" className={classes.field} required>
                    <InputLabel className={classes.label} id="averageDurationLabel">
                        Durata media del servizio
                    </InputLabel>
                    <Select
                        labelId="averageDurationLabel"
                        name="averageDuration"
                        inputProps={{ 'data-testid': 'averageDuration' }}
                        defaultValue={service.averageDuration}
                    >
                        {PROGRESSIVE_MINS_WITHIN_MAX_SERVICE.map((valueInMins) => (
                            <MenuItem key={valueInMins} value={valueInMins}>
                                {formatMinutes(valueInMins)}
                            </MenuItem>
                        ))}
                    </Select>
                </FormControl>

                <div className={classes.submitButton}>
                    <Button
                        type="submit"
                        color="primary"
                        variant="contained"
                        disabled={loading}
                        startIcon={loading ? <CircularProgress size={20} /> : null}
                        data-testid="submit-button"
                    >
                        {loading ? 'Caricamento' : 'Applica modifiche'}
                    </Button>
                    <Button variant="contained" color="primary" disabled={loading} onClick={() => history.goBack()}>
                        Annulla
                    </Button>
                </div>
            </form>
        </div>
    );
};

export default React.memo(
    AddService,
    (prevProps, nextProps) => prevProps.location.pathname === nextProps.location.pathname,
);
