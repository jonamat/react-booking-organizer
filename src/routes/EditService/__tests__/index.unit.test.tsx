/**
 * Unit test suite
 *
 * Target: EditService component
 * Tests:
 * - Responses handling
 * - Redirection
 * - User interaction interruption on loading
 * - Middleware function call
 * - Visual feedbacks
 * - Snapshot matching
 * - Functionality
 *
 * For data validation, API call and errors handling see {@link '../../../utils/editService'}
 */

// Test resources
import React from 'react';
import renderer from 'react-test-renderer';
import { StylesProvider, createGenerateClassName } from '@material-ui/core';
import { render, fireEvent, waitFor } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';

// Test configuration
import { editService } from '../../../utils';
import { __mockServices } from '../../../__mocks__/react-redux';

// Mocked modules
import { toast } from 'react-toastify';

// Component to test
import EditService from '..';

const match = {
    params: {
        serviceId: __mockServices[0].id,
    },
};

/* -------------------------------------------------------------------------- */
/*                                    Mocks                                   */
/* -------------------------------------------------------------------------- */

jest.mock('../../../utils', () => ({
    ...(jest.requireActual('../../../utils') as Record<string, any>),
    editService: jest.fn(() => Promise.resolve({ id: 'fakeID', name: 'Carl' })),
}));
const spyTSuccess = jest.spyOn(toast, 'success');
const spyTError = jest.spyOn(toast, 'error');
const history = {
    push: jest.fn(),
};

/* -------------------------------------------------------------------------- */
/*                              Bridge operations                             */
/* -------------------------------------------------------------------------- */

beforeEach(() => {
    jest.clearAllMocks();
});

/* -------------------------------------------------------------------------- */
/*                               Render wrapper                               */
/* -------------------------------------------------------------------------- */

/**
 * Render the component with the custom configuration
 * @param props Component props
 */
const renderComponent = (props?) => {
    return render(<EditService history={history} match={match} {...props} />);
};

/* -------------------------------------------------------------------------- */
/*                                    Tests                                   */
/* -------------------------------------------------------------------------- */

describe('EditService component', () => {
    describe('Upload system', () => {
        it('calls the db middleware function with correct user inputs', () => {
            const { getByTestId } = renderComponent();

            fireEvent.input(getByTestId('name'), { target: { value: 'name' } });
            fireEvent.input(getByTestId('averageDuration'), { target: { value: 30 } });
            fireEvent.click(getByTestId('submit-button'));

            expect(spyTError).not.toBeCalled();
            expect(editService as jest.Mock).toBeCalledTimes(1);
            expect((editService as jest.Mock).mock.calls[0][0]).toEqual(__mockServices[0].id);
            expect(Array.from((editService as jest.Mock).mock.calls[0][1].entries())).toEqual([
                ['name', 'name'],
                ['averageDuration', '30'],
            ]);
        });
    });

    describe('User experience', () => {
        it('redirects user on success', async () => {
            const { getByTestId } = renderComponent();

            fireEvent.input(getByTestId('name'), { target: { value: 'name' } });
            fireEvent.input(getByTestId('averageDuration'), { target: { value: 15 } });
            fireEvent.click(getByTestId('submit-button'));

            await waitFor(() => {
                expect(history.push).toBeCalledTimes(1);
                expect(history.push).toBeCalledWith('/services');
            });
        });

        it('calls toast on success', async () => {
            const { getByTestId } = renderComponent();

            fireEvent.input(getByTestId('name'), { target: { value: 'name' } });
            fireEvent.input(getByTestId('averageDuration'), { target: { value: 15 } });
            fireEvent.click(getByTestId('submit-button'));

            await waitFor(() => {
                expect(spyTSuccess).toBeCalledTimes(1);
            });
        });

        it('calls toast on error', async () => {
            (editService as jest.Mock).mockImplementation(() => Promise.reject('some error'));

            const { getByTestId } = renderComponent();

            fireEvent.input(getByTestId('name'), { target: { value: 'name' } });
            fireEvent.input(getByTestId('averageDuration'), { target: { value: 15 } });
            fireEvent.click(getByTestId('submit-button'));

            await waitFor(() => {
                expect(spyTError).toBeCalledTimes(1);
            });
        });

        it('disables submit button on loading', async () => {
            (editService as jest.Mock).mockImplementation(() => Promise.reject('some error'));

            const { getByTestId } = renderComponent();

            fireEvent.input(getByTestId('name'), { target: { value: 'name' } });
            fireEvent.input(getByTestId('averageDuration'), { target: { value: 15 } });

            expect(getByTestId('submit-button').hasAttribute('disabled')).toBeFalsy();
            fireEvent.click(getByTestId('submit-button'));
            expect(getByTestId('submit-button').hasAttribute('disabled')).toBeTruthy();

            await waitFor(() => {
                expect(getByTestId('submit-button').hasAttribute('disabled')).toBeFalsy();
            });
        });
    });

    it('matches snapshot', () => {
        const snapshot = renderer
            .create(
                <StylesProvider generateClassName={createGenerateClassName({ disableGlobal: true })}>
                    {/* @ts-ignore */}
                    <EditService match={match} />
                </StylesProvider>,
            )
            .toJSON();
        expect(snapshot).toMatchSnapshot();
    });
});
