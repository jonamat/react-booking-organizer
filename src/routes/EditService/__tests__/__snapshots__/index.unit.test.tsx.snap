// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditService component matches snapshot 1`] = `
<div
  className="jss1"
  data-testid="edit-service"
>
  <h6
    className="jss7 jss17"
  >
    Aggiungi servizio
  </h6>
  <hr
    className="jss37 jss2"
  />
  <form
    className="jss3"
    data-testid="form"
    onSubmit={[Function]}
  >
    <div
      className="jss45 jss44 jss4"
    >
      <label
        className="jss61 jss49 jss55 jss58 jss57 jss60 jss64 jss51 jss66 jss67 jss53"
        data-shrink={true}
      >
        ID
        <span
          aria-hidden={true}
          className="jss68 jss54"
        >
           
          *
        </span>
      </label>
      <div
        className="jss84 jss69 jss87 jss72 jss85"
        onClick={[Function]}
      >
        <input
          aria-invalid={false}
          autoComplete="off"
          autoFocus={false}
          className="jss95 jss79 jss87 jss72"
          defaultValue="service-FAKE-ID1"
          disabled={true}
          name="id"
          onAnimationStart={[Function]}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          required={true}
          type="text"
        />
        <fieldset
          aria-hidden={true}
          className="jss102 jss78"
        >
          <legend
            className="jss104 jss105"
          >
            <span>
              ID
               *
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
    <div
      className="jss45 jss44 jss4"
    >
      <label
        className="jss61 jss49 jss55 jss58 jss57 jss60 jss66 jss67 jss53"
        data-shrink={true}
      >
        Nome
        <span
          aria-hidden={true}
          className="jss68 jss54"
        >
           
          *
        </span>
      </label>
      <div
        className="jss84 jss69 jss85"
        onClick={[Function]}
      >
        <input
          aria-invalid={false}
          autoComplete="off"
          autoFocus={false}
          className="jss95 jss79"
          data-testid="name"
          defaultValue="fake service 1"
          disabled={false}
          name="name"
          onAnimationStart={[Function]}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          required={true}
          type="text"
        />
        <fieldset
          aria-hidden={true}
          className="jss102 jss78"
        >
          <legend
            className="jss104 jss105"
          >
            <span>
              Nome
               *
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
    <div
      className="jss45 jss4"
    >
      <label
        className="jss61 jss49 jss5 jss55 jss58 jss57 jss60 jss66 jss67 jss53"
        data-shrink={true}
        id="averageDurationLabel"
      >
        Durata media del servizio
        <span
          aria-hidden={true}
          className="jss68 jss54"
        >
           
          *
        </span>
      </label>
      <div
        className="jss84 jss69 jss85"
        onClick={[Function]}
      >
        <div
          aria-haspopup="listbox"
          aria-labelledby="averageDurationLabel mui-component-select-averageDuration"
          className="jss106 jss107 jss110 jss109 jss95 jss79"
          id="mui-component-select-averageDuration"
          onBlur={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onMouseDown={[Function]}
          role="button"
          tabIndex={0}
        >
          30 minuti
        </div>
        <input
          aria-hidden={true}
          className="jss116"
          data-testid="averageDuration"
          name="averageDuration"
          onAnimationStart={[Function]}
          onChange={[Function]}
          required={true}
          tabIndex={-1}
          value={30}
        />
        <svg
          aria-hidden={true}
          className="jss117 jss112 jss115"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M7 10l5 5 5-5z"
          />
        </svg>
        <fieldset
          aria-hidden={true}
          className="jss102 jss78"
          style={
            Object {
              "paddingLeft": 8,
            }
          }
        >
          <legend
            className="jss103"
            style={
              Object {
                "width": 0.01,
              }
            }
          >
            <span
              dangerouslySetInnerHTML={
                Object {
                  "__html": "&#8203;",
                }
              }
            />
          </legend>
        </fieldset>
      </div>
    </div>
    <div
      className="jss6"
    >
      <button
        className="jss159 jss130 jss138 jss139"
        data-testid="submit-button"
        disabled={false}
        onBlur={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="submit"
      >
        <span
          className="jss131"
        >
          Applica modifiche
        </span>
      </button>
      <button
        className="jss159 jss130 jss138 jss139"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <span
          className="jss131"
        >
          Annulla
        </span>
      </button>
    </div>
  </form>
</div>
`;
